examples/BERT/model.py

```diff
@@ -171,21 +176,137 @@ class CrossLingualMLMTask(nn.Module):
+
         self.mlm_span = Linear(ninp, ninp)
         self.activation = F.gelu
         self.norm_layer = LayerNorm(ninp, eps=1e-12)
         self.mlm_head = Linear(ninp, ntoken)
 
     def forward(self, src):
-        output = self.xlmr_model(src)
+        # output = self.xlmr_model(src)
+        output = self.xlmr_embed(src)
+        output = self.transformer_encoder1(output)
+        # ======================== Split ==========================
+        output = self.transformer_encoder2(output)
+
         output = self.mlm_span(output)
         output = self.activation(output)
         output = self.norm_layer(output)
         output = self.mlm_head(output)
         return output
 
 
+def get_cuda_if_available(i):
+    assert i >= 0
+    if torch.cuda.is_available():
+        return f"cuda:{min(i, torch.cuda.device_count() - 1)}"
+    else:
+        return "cpu"
+
+
+class CrossLingualMLMTaskBase(nn.Module):
+    def __init__(self, device):
+        super(CrossLingualMLMTaskBase, self).__init__()
+        self.device = device
+        self._lock = threading.Lock()
+
+    def forward(self, x_rref):
+        x = x_rref.to_here().to(self.device)
+        with self._lock:
+            out = self._forward(x)
+        return out.cpu()
+
+    def parameter_rrefs(self):
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [RRef(p) for p in self.parameters()]
+
+
+class CrossLingualMLMTaskShard1(CrossLingualMLMTaskBase):
+    def __init__(self, device, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5):
+        super(CrossLingualMLMTaskShard1, self).__init__(device)
+        self.xlmr_embed = XLMREmbedding(ntoken, ninp, dropout).to(device)
+        encoder_layers = TransformerEncoderLayer(ninp, nhead, nhid, dropout)
+        self.transformer_encoder = TransformerEncoder(encoder_layers, nlayers // 2).to(device)
+
+    def _forward(self, src):
+        output = self.xlmr_embed(src)
+        output = self.transformer_encoder(output)
+        return output
+
+
+class CrossLingualMLMTaskShard2(CrossLingualMLMTaskBase):
+    def __init__(self, device, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5):
+        super(CrossLingualMLMTaskShard2, self).__init__(device)
+        encoder_layers = TransformerEncoderLayer(ninp, nhead, nhid, dropout)
+        self.transformer_encoder = TransformerEncoder(encoder_layers, nlayers // 2).to(device)
+        self.mlm_span = Linear(ninp, ninp).to(device)
+        self.activation = F.gelu
+        self.norm_layer = LayerNorm(ninp, eps=1e-12).to(device)
+        self.mlm_head = Linear(ninp, ntoken).to(device)
+
+    def _forward(self, src):
+        output = self.transformer_encoder(src)
+        output = self.mlm_span(output)
+        output = self.activation(output)
+        output = self.norm_layer(output)
+        output = self.mlm_head(output)
+        return output
+
+
+class DistCrossLingualMLMTask(nn.Module):
```

Comment:
Seem to me that this could be a container which users can pass model shards, workers, arguments to. And the container will handle the forward func (a.k.a. splitting input batch, RPC reference, and concatenating the output.

```
class RPCShardContainer(nn.Module):
    def __init__(self, split_size, model_shard, workers, *args, **kwargs):
        ...
    def forward(self, data_batch):
        ...
model = RPCShardContainer(2, devices, [CrossLingualMLMTaskShard1, CrossLingualMLMTaskShard2], ...)
```